Fixes #3383 - create about_calculated_properties (#6455)
MicrosoftDocs/PowerShell-Docs: reference/7.0/Microsoft.PowerShell.Core/About/about_Calculated_Properties.md

---
Locale: en-US
ms.date: 08/07/2020
online version: https://docs.microsoft.com/powershell/module/microsoft.powershell.core/about/about_calculated_properties?view=powershell-7&WT.mc_id=ps-gethelp
schema: 2.0.0
title: about_Calculated_Properties
---
# About Calculated Properties

## Short Description

> Calculated properties allow you to dynamically add properties and formatting to

Author: nope

objects in the pipeline.

## Long Description

> PowerShell provides several cmdlets that operate on properties of objects.

Reviewer: This reads kind of strangely as written. 
- Two consecutive sentences that start the exact same way:
  - "PowerShell provides...PowerShell Provides..."
- The first sentence as written, seems like subject is about cmdlets that operate on properties and not necessarily speaking specifically about Calculated Properties.This entire section could probably be written something like this:

PowerShell provides several cmdlets that can operate on properties of objects using a method to create Calculated Properties. Calculated Properties are used to dynamically create new properties that are added to the output object or provide custom formatting for an existing property.

I removed the sentence, "Calculated Properties are dynamically created at run time" because as written, it doesn't really add any useful information and is more confusing than anything else. If It would remove critical information to take it out, it should be clarified and simplified to express that information. Or, be followed by a clarifghying sentence that adds more info?

Author: I reworked that paragraph

> PowerShell provide a method to create calculated properties. Calculated

Reviewer 2: provides

Author: Fixed

> properties are dynamically defined at runtime. Calculated properties can be use

Reviewer 3: ```suggestion
properties are dynamically defined at formatting time. Calculated properties can be use
```

> properties are dynamically defined at runtime. Calculated properties can be use

Reviewer 2: used

Author: Reworked this paragraph

to dynamically create new properties that are added to the output object or
provide custom formatting for an existing property.

> Each cmdlet that supports calculated properties has specific set of attributes

Reviewer 2: has a specific

Author: fixed

that describe the property. The attributes define the name of the property, an
expression to calculate the value, and optional formatting information. A
calculated property is defined using a hashtable containing these attributes as
key-value pairs.

## Supported cmdlets

The following cmdlets support calculated properties using the listed property
attributes names as keys in the hashtable.

> [!NOTE]
> The `expression` attribute can be a script block instead of a hashtable. For
> more information, see the [Notes](#notes) section. For descriptions of these
> attibutes, see [Attribute definitions](#attribute-definitions).

- `Compare-Object`
  - `expression`

- `ConvertTo-Html`
  - `name`/`label` - optional
  - `expression`
  - `width` - optional
  - `alignment` - optional

- `Format-Custom`
  - `expression`
  - `depth` - optional

- `Format-List`
  - `name`/`label` - optional
  - `expression`
  - `formatstring` - optional

>   This same set of attributes also apply to the `-GroupBy` properties for all

Reviewer 2: applies

Reviewer: I don't think this should be applies? I could be wrong but I think apply is _applied_ correctly ;)

Author: It is applies - "set" is a singular noun.

  `Format-*` cmdlets.

- `Format-Table`
  - `name`/`label` - optional
  - `expression`
  - `formatstring` - optional
  - `width` - optional
  - `alignment` - optional

- `Format-Wide`
  - `expression`
  - `formatstring` - optional

- `Group-Object`
  - `expression`

- `Measure-Object`
  - Only supports a script block for the expression, not a hashtable.

- `Select-Object`
  - `name`/`label` - optional
  - `expression`

- `Sort-Object`
  - `expression`
  - `ascending`/`descending` - optional

For more information, see the [Examples](#examples) section of this article.

## Attribute definitions

- `name`/`label` - Specifies the name of the property being created. You can
  use `name` or its alias, `label`, interchangeably.
- `expression` - A script block used to calculate the value of the new property.
- `alignment` - Used by cmdlets that produce tabular output to define how the

>   values is displayed in a column. The value must be `'left'`, `'center'`, or

Reviewer 2: Should these values be both codefenced and surrounded by ticks?

Author: They must be strings. They are not enums. So this is correct.

>   values is displayed in a column. The value must be `'left'`, `'center'`, or

Reviewer: ```suggestion
  values are displayed in a column. The value must be `'left'`, `'center'`, or
```

Author: Fixed

  `'right'`.
- `formatstring` - Specifies a format string that defines how the value is
  formatted for output. For more information about format strings, see
  [Format types in .NET](/dotnet/standard/base-types/formatting-types).
- `width` - Specifies the maximum width column in a table when the value is
  displayed. The value must be greater than `0`.
- `depth` - The **Depth** parameter of `Format-Custom` specifies the depth of
  expansion for all properties. The `depth` attribute allows you to specify the
  depth of expansion per property.
- `ascending` / `descending` - Allows you to specify the order of sorting for
  one or more properties. These are boolean values.

The hashtable keys need not be spelled out as long as the specified name prefix
is unambiguous. For example, `n` can be used in lieu of `Name` and `e` can be

> used in lieu of `Expression`.

Reviewer: could a negative example be shown to highlight the concept in both forms?

Author: The examples use a variety of forms.

## Examples

### Compare-Object

With calculated properties, you can control how the properties of the input
objects are compared.

> 

Reviewer: could add a sentence clarifying the example a little.?

Author: Added.

```powershell
Compare-Object @{p=1} @{p=2} -property @{ Expression = { $_.p % 2 } }
```

```Output
 $_.p % 2  SideIndicator
---------- -------------
         0 =>
         1 <=
```

### ConvertTo-Html

`ConvertTo-Html` can convert a collection of objects to an HTML table.
Calculated properties allow you to control how the table is presented.

```powershell
Get-Alias |
  ConvertTo-Html Name,
                 Definition,
                 @{n='ParamCount';e={$_.Parameters.Keys.Count};align='center'} |
    Out-File .\aliases.htm -Force
```

> This example creates an HTML table contain a list of PowerShell aliases and the

Reviewer 2: containing or that contains

Author: Fixed

number parameters for each aliased command. The values of **ParamCount** column
are centered.

### Format-Custom

`Format-Custom` provides a custom view of an object in a format similar to a
class definition. More complex objects can contain members that are deeply
nested with complex types. The **Depth** parameter of `Format-Custom` specifies
the depth of expansion for all properties. The `depth` attribute allows you to
specify the depth of expansion per property.

> In this example we use the `depth` attribute to simplify the custom output for

Reviewer 2: In this example,

Author: Fixed

> the `Get-Date` cmdlet . `Get-Date` returns a **DateTime** object. The **Date**

Reviewer 2: bonus space after cmdlet

Author: fixed

property of this object is also a **DateTime** object, so the object is nested.

```powershell

> Get-Date | Format-Custom @{e={$_.Date};depth=1},TimeOfDay

Reviewer: I know it adds some length but we should avoid aliases in code samples as much as possible?

Author: I chose to use short form to two reasons:
1. to demonstrate the short form
2. limited column width of About_ topics

```

```Output
class DateTime
{
  $_.Date =
    class DateTime
    {
      Date = 8/7/2020 12:00:00 AM
      Day = 7
      DayOfWeek = Friday
      DayOfYear = 220
      Hour = 0
      Kind = Local
      Millisecond = 0
      Minute = 0
      Month = 8
      Second = 0
      Ticks = 637323552000000000
      TimeOfDay = 00:00:00
      Year = 2020
      DateTime = Friday, August 07, 2020 12:00:00 AM
    }
  TimeOfDay =
    class TimeSpan
    {
      Ticks = 435031592302
      Days = 0
      Hours = 12
      Milliseconds = 159
      Minutes = 5
      Seconds = 3
      TotalDays = 0.503508787386574
      TotalHours = 12.0842108972778
      TotalMilliseconds = 43503159.2302
      TotalMinutes = 725.052653836667
      TotalSeconds = 43503.1592302
    }
}
```

### Format-List

In this example, we use calculated properties to change the name and format of
the output from `Get-ChildItem`.

```powershell
Get-ChildItem ..\*.json -File |
  Format-List Fullname,
              @{
                 name='Modified'

>                  expr={$_.LastWriteTime}

Reviewer: Seems inconsistent to have `expression` written this way.

Author: Fixed

        switch ($_.extension) {
          '.md'   {'Content'}
          ''      {'Metacontent'}
          '.ps1'  {'Automation'}
          '.yml'  {'Automation'}
          default {'Configuration'}
        }
      }
    }
```

```Output
   Type: Metacontent

Name              Length
----              ------
ThirdPartyNotices   1229
LICENSE-CODE        1106
LICENSE            19047

   Type: Configuration

Name                                Length
----                                ------
.editorconfig                          183
.gitattributes                         419
.gitignore                             228
.markdownlint.json                    2456
.openpublishing.publish.config.json   2306
.openpublishing.redirection.json    332394
.localization-config                   232

   Type: Content

Name            Length
----            ------
README.md         3355
CONTRIBUTING.md    247

   Type: Automation

Name                      Length
----                      ------
.openpublishing.build.ps1    796
build.ps1                   7495
ci.yml                       645
ci-steps.yml                2035
daily.yml                   1271
```

### Format-Wide

> The `Format-Wide` cmdlet allows you to display the value one property for

Reviewer 2: value of one property

Author: Fixed

objects in a collection as a multi-column list.

For this example, we want to see the filename and the size (in kilobytes) as a
wide listing. Since `Format-Wide` does not display more than one property, we
use a calculated property to combine the value of two properties into a single
value.

```powershell
Get-ChildItem -File |
  Format-Wide -Property @{e={'{0} ({1:N2}kb)' -f $_.name,($_.length/1kb)}}
```

```Output
.editorconfig (0.18kb)                          .gitattributes (0.41kb)
.gitignore (0.22kb)                             .localization-config (0.23kb)
.markdownlint.json (2.40kb)                     .openpublishing.build.ps1 (0.78kb)
.openpublishing.publish.config.json (2.25kb)    .openpublishing.redirection.json (324.60kb)
build.ps1 (7.32kb)                              ci.yml (0.63kb)
ci-steps.yml (1.99kb)                           CONTRIBUTING.md (0.24kb)
daily.yml (1.24kb)                              LICENSE (18.60kb)
LICENSE-CODE (1.08kb)                           README.md (3.28kb)
ThirdPartyNotices (1.20kb)
```

### Group-Object

The `Group-Object` cmdlet displays objects in groups based on the value of a

> specified property. In this example we are using a calculated property to count

Reviewer 2: In this example,

Author: Fixed

the number of files of each content type.

```powershell
Get-ChildItem -File |
  Sort-Object extension |
    Group-Object -NoElement -Property @{
      name='Type'
      expression={
        switch ($_.extension) {
          '.md'   {'Content'}
          ''      {'Metacontent'}
          '.ps1'  {'Automation'}
          '.yml'  {'Automation'}
          default {'Configuration'}
        }
      }
    }
```

```Output
Count Name
----- ----
    5 Automation
    7 Configuration
    2 Content
    3 Metacontent
```

### Measure-Object

The `Measure-Object` cmdlet calculates the numeric properties of objects. In
this example, we use a calculated property to get the count (**Sum**) of the
numbers, between 1 and 10, that are evenly divisible by 3.

```powershell
1..10 | Measure-Object -Property {($_ % 3) -eq 0} -Sum
```

```Output
Count             : 10
Average           :
Sum               : 3
Maximum           :
Minimum           :
StandardDeviation :
Property          : ($_ % 3) -eq 0
```

> [!NOTE]
> Unlike the other cmdlets, `Measure-Object` does not accept a hashtable for
> calculated properties. You must only use a script block.

### Select-Object

> You can use calculated properties to add additional members to the objects

Reviewer: ```suggestion
You can use calculated properties to add additional members to the object's
```

Author: It is not a possessive case. I am talking about the objects being output. Not the output from the object.

> output by the `Select-Object` cmdlet. In this example, we are listing the

Reviewer 2: with the `Select-Object` cmdlet.

Author: Fixed

GetHashCode    Method       int GetHashCode()
GetType        Method       type GetType()
ToString       Method       string ToString()
Definition     NoteProperty string Definition=Get-Content
Name           NoteProperty string Name=cat
ParameterCount NoteProperty System.Int32 ParameterCount=21

Name    Definition         ParameterCount
----    ----------         --------------
cat     Get-Content                    21
cd      Set-Location                   15
cdd     Push-MyLocation                 1
chdir   Set-Location                   15
clc     Clear-Content                  20
clear   Clear-Host                      0
clhy    Clear-History                  17
cli     Clear-Item                     20
clp     Clear-ItemProperty             22
cls     Clear-Host                      0
clv     Clear-Variable                 19
cnsn    Connect-PSSession              29
compare Compare-Object                 20
copy    Copy-Item                      24
cp      Copy-Item                      24
cpi     Copy-Item                      24
cpp     Copy-ItemProperty              23
cvpa    Convert-Path                   13
```

### Sort-Object

> Using use calculated properties, you can sort data in different orders per

Reviewer 2: Using the calculated

Author: Fixed

property. This example sorts data from a CSV file in ascending order by
**Date**. But within each date, we are sorting the rows in descending order by
**UnitsSold**.

```powershell
Import-Csv C:\temp\sales-data.csv |
  Sort-Object Date, @{expr={$_.UnitsSold}; desc=$true}, Salesperson  |
    Select-Object Date, Salesperson, UnitsSold
```

```Output
Date       Salesperson UnitsSold
----       ----------- ---------
2020-08-01 Sally       3
2020-08-01 Anne        2
2020-08-01 Fred        1
2020-08-02 Anne        6
2020-08-02 Fred        2
2020-08-02 Sally       0
2020-08-03 Anne        5
2020-08-03 Sally       3
2020-08-03 Fred        1
2020-08-04 Anne        2
2020-08-04 Fred        2
2020-08-04 Sally       2
```

## Notes

- You may specify the expression script block _directly_, as an argument,
  rather than specifying it as the `Expression` entry in a hashtable. For
  example:

  ```powershell
  '1', '10', '2' | Sort-Object { [int] $_ }
  ```

>   This is convenient for cmdlets that do not require (or support) naming a

Reviewer 2: This example

Author: Fixed

  property via the `Name` key, such as `Sort-Object`, `Group-Object` and
  `Measure-Object`.

  For cmdlets that support naming the property, the script block is converted
  to a string and used as the name of the property in the output.

> - Expression script blocks run in _child_ scopes, meaning that the caller's

Reviewer: Should expression be code fenced?

Author: Fixed

---
Locale: en-US
ms.date: 08/07/2020
online version: https://docs.microsoft.com/powershell/module/microsoft.powershell.core/about/about_calculated_properties?view=powershell-7&WT.mc_id=ps-gethelp
schema: 2.0.0
title: about_Calculated_Properties
---
# About calculated properties

## Short Description

Calculated properties allow you to dynamically add properties and formatting to
objects in the pipeline.

## Long Description

There are several cmdlets that operate on properties of objects. PowerShell
provides a method to create calculated properties. Calculated properties
dynamically create new properties that are added to the output object and can
provide custom formatting for an existing property.

Each cmdlet that supports calculated properties has a specific set of
attributes that describe the property. The attributes define the name of the
property, an expression to calculate the value, and optional formatting
information. A calculated property is defined using a
[hashtable](about_hash_tables.md) containing these attributes as key-value

> pairs.

Reviewer 4: I think approaching this subject from the angle of the properties on the output objects themselves confuses the material. You can see that in this first paragraph where we're forced to overload "properties" to mean both the parameter and the property on the output object.

My feeling is that it would be better to write the article *about the common parameters* that generate calculated properties, rather than trying to write directly about the properties themselves.

Otherwise:

> There are several cmdlets that operate on properties of objects.

This phrase doesn't tell me anything about what you want me to know, and is too general for me to really understand what you're getting at.

> PowerShell provides a method to create calculated properties.

This doesn't relate strongly to the previous statement, refers to "a method" without qualifying it in any way, and uses "calculated properties" as the object of the sentence without introducing or defining the concept.

> Calculated properties dynamically create new properties that are added to the output object and can provide custom formatting for an existing property.

The word properties is used twice here to mean different things, we depend on the also unintroduced concept of the "output object" and within the same sentence launch into the subject of formatting.

> Each cmdlet that supports calculated properties has a specific set of
attributes that describe the property.

We're now back to cmdlets, from objects and output objects, but refer to "the property" without the definiteness the article implies. Plus we've introduced attributes.

And so on.

Instead, addressing the subject from the entry point of the parameter:

> A number of PowerShell cmdlets that transform, aggregate or process input objects into output objects provide parameters that allow the addition of new properties to those output objects.
> These property-adding parameters can be used to generate new *calculated properties* on output objects based on the values of input objects.
> In particular, a property-adding parameter takes a hashtable value which can be used to set the name, calculating scriptblock expression and formatting information for a property it generates on an output object.

Author: @rjmholt I reworked the intro. Let me know if that reads better.
